<img src=https://www.meme-arsenal.com/memes/5f0ccdd82eda80ded9db721a8746f804.jpg><br>
<hr style="border: 10px solid green;" />

## <span style="color:green">Финальный комментарий<span>
Спасибо за проделанную работу🤝

Камиля, отлично, ты нашла дубликаты и устранила их ✅ Дальше будет больше интересного, желаю успехов в учебе 🖐

Поздравляю, проект принят👍

<hr style="border: 2px solid green;" />

## Комментарий ревьювера 2
Камиля, привет!

Спасибо за доработки, очень классно, что ты воспользовалась рекомендациями.

Оставил [комментарий](#21-bullet) по дубликатам

<hr style="border: 2px solid green;" />

## Комментарий ревьювера

Камиля, привет! 

🤚Меня зовут, Олег Юрьев, я буду проверять твой проект. Если не против, я буду обращаться к тебе на _ТЫ_ . Если это неприемлемо, обязательно напиши мне в комментариях - мы перейдем на _ВЫ_ .

🤝Спасибо, что сдала проект вовремя. Тобой проделана большая работа, молодец! 

✏️В этой тетрадке ты найдешь мои комментарии.

Комментарии будут выделены:

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
    Безупречная работа</div>       
<br>
<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
    Ничего дорабатывать не надо, данные пункты работают нормально, но есть способы улучшить. Несколько таких комментариев допустимы, но более - придется некоторые доработать. Большое кол-во мелких замечаний и помарок сказываются на общем впечатлении от работы.</div>   
<br>
<div style="border:solid red 2px; padding: 20px"> <b>Что необходимо доработать</b><br>
    Такие пункты необходимо исправить в первую очередь, чтобы я смог принять проект.</div>
 
Не удаляй комментарии, так будем следить за динамикой 📈
   
Давай будем в диалоге - основные цели в диалоге:
- обмен опытом (да я тоже учусь у студентов👨‍🎓)
- выполнить проект еще лучше
- помочь тебе стать аналитиком

Поэтому чтобы мне проще было проверять повторно, а тебе контролировать процесс ревью, оставляй свои комментарии и выделяй их. 
    
<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Комментарий студента:</b> Например, вот так.</div>
   
Поехали! :)

<hr style="border: 2px solid green;" />

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
ОЧень здорово, что ты оставляешь вводную информацию. Твои коллеги, начальник сразу поймут о чем пойдет речь!</div> <br>
<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Рекомендую, более строго формализовать цель исследования</div> <br>

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('/datasets/data.csv')

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Рекомендую, разделять логически разный код:
    <ul>
    <li> импорты библиотек
    <li> открытие данных
    <li> рабочий код</ul>
Такой подход улучшить читаемость и уменьшит время на поиск ошибок (если они есть)</div> <br>

Прочитали файл *data.csv*, сохранили его в переменной *data* и получили общую информацию о данных. Всего в таблице 12 столбцов с типами данных: *object*, *float* и *int*. 

Подробная информация о столбцах: 
* children — количество детей в семье;
* days_employed — общий трудовой стаж в днях;
* dob_years — возраст клиента в годах;
* education — уровень образования клиента;
* education_id — идентификатор уровня образования;
* family_status — семейное положение;
* family_status_id — идентификатор семейного положения;
* gender — пол клиента;
* income_type — тип занятости;
* debt — имел ли задолженность по возврату кредитов;
* total_income — ежемесячный доход;
* purpose — цель получения кредита.

В столбцах *days_employed* и *total_income* количество значений отличается от других, значит, в данных имеются пропущенные значения. 

In [4]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [5]:
data_copy = data.copy(deep=True)

**Вывод**

Из таблицы мы видим, что в столбце *days_employed* присутствуют отрицательные значения, что является артефактом и не соответствует действительности. При выгрузке данных могла произойти ошибка, из-за чего значения стали отрицательными. Также в четвертой строке этого столбца можно заметить, что трудовой стаж равен *340266 дням*, что соответствует *932 годам*. Данные ошибки нужно будет отработать и исправить. 

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
Отлично, обзор данных выполнен хорошо, здорово, что ты отметила найденные инсайты в выводе</div> <br>

## Шаг 2. Предобработка данных

### Обработка пропусков

In [6]:
data.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

По суммарному количеству пропусков видно, что в некоторых строках пропущена информация о трудовом стаже и ежемесячном доходе. Это можно объяснить тем, что некоторые люди не работают по разным причинам, следовательно не имеют дохода. Данные пропуски необходимо заменить средними значениями по столбцу. Для этого сначала нужно получить абсолютные значения столбца *days_employed*, проверить строки с трудовым стажем, равным сотням лет и заменить их для корректного расчета среднего значения.

In [7]:
data['days_employed'] = data['days_employed'].abs() #нашли абсолютные значения отрицательного стажа
data.head(5)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [8]:
data['dob_years'].max() #максимальный возраст

75

75 лет - максимальный возраст из столбца *dob_years*. Тогда его максимальный рабочий стаж будет равен *250 рабочих дней в год * (75-16) лет = 14750 дней*. Возьмем это значение за максимально возможное в этом столбце. 

In [9]:
data[data['days_employed'] > 14750]

children  days_employed  dob_years education  education_id  \
4             0  340266.072047         53   среднее             1   
18            0  400281.136913         53   среднее             1   
24            1  338551.952911         57   среднее             1   
25            0  363548.489348         67   среднее             1   
30            1  335581.668515         62   среднее             1   
...         ...            ...        ...       ...           ...   
21505         0  338904.866406         53   среднее             1   
21508         0  386497.714078         62   среднее             1   
21509         0  362161.054124         59    высшее             0   
21518         0  373995.710838         59   СРЕДНЕЕ             1   
21521         0  343937.404131         67   среднее             1   

               family_status  family_status_id gender income_type  debt  \
4           гражданский брак                 1      F   пенсионер     0   
18            вдовец / вдова                 2      F   пенсионер     0   
24     Не женат / не замужем                 4      F   пенсионер     0   
25           женат / замужем                 0      M   пенсионер     0   
30           женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
21505       гражданский брак                 1      M   пенсионер     0   
21508        женат / замужем                 0      M   пенсионер     0   
21509        женат / замужем                 0      M   пенсионер     0   
21518        женат / замужем                 0      F   пенсионер     0   
21521        женат / замужем                 0      F   пенсионер     0   

        total_income                                purpose  
4      158616.077870                        сыграть свадьбу  
18      56823.777243     на покупку подержанного автомобиля  
24     290547.235997  операции с коммерческой недвижимостью  
25      55112.757732                   покупка недвижимости  
30     171456.067993  операции с коммерческой недвижимостью  
...              ...                                    ...  
21505   75439.993167                        сыграть свадьбу  
21508   72638.590915                           недвижимость  
21509   73029.059379               операции с недвижимостью  
21518  153864.650328                   сделка с автомобилем  
21521  155999.806512                   сделка с автомобилем  

[3459 rows x 12 columns]

Проверяем артефакты: 14750 дней - максимально возможный официальный стаж работы (250 рабочих дней в году) с 16 лет до максимального возраста в таблице. Из полученной отфильтрованной таблицы видно, что все эти люди - пенсионеры, поэтому для более аккуратного результата применим корреляцию данных. Найдем отношение медианы работников с огромным стажем к медиане с естественным стажем, затем уменьшим "тысячелетние" стажи на этот коэффициент и таким образом проведем нормирование данных.

In [10]:
med_over = data[data['days_employed'] > 14750]['days_employed'].median()
med_below = data[data['days_employed'] < 14750]['days_employed'].median()
coef = med_over/med_below
data.loc[data['days_employed'] > 14750,'days_employed'] = data[data['days_employed'] > 14750]['days_employed']/coef
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0    1518.332393         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Теперь найдем среднее значение в этом столбце и заменим им пропуски:

In [11]:
days_employed_avg = data['days_employed'].mean()
data['days_employed'] = data['days_employed'].fillna(value=days_employed_avg)
data.isnull().sum()

children               0
days_employed          0
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Теперь пропуски остались в столбце *total_income*, заменим их. Так как доход может сильно различаться по столбцу *income_type*, то есть по типу заработка, отработаем пропуски через замену на медиану данного вида заработка с помощью функции transform(), которая в данном случае принимает пустое значение ячейки и заменяет его на медиану данной категории.

In [12]:
data['total_income'] = data.groupby('income_type')['total_income'].transform(lambda x: x.fillna(value=x.median()))
data.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Все пропуски отработаны, их больше не осталось.

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
Интересный способ обработки некорректных значений признака <code>days_employed</code>. Единственное, я бы не стал трогать исходные данные, а создал бы отдельный столбец.</div> <br>
<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Кажется, что заполнение средним не самый лучший вариант, подумай может использовать какую-либо группировку при заполнении?! Например, общий доход может сильно разнится между студентом и служащим или пенсионером, а так ты всех подводишь под 1 гребенку</div> <br>

In [13]:
# рассчитаем средний общий доход по группам, кажется, что различия есть и они существенные
data_copy\
    .groupby('income_type')\
    .agg({'total_income': 'mean'})\
    .sort_values(by='total_income', ascending=False)\
    .style.bar(color='#3690FF')

<hr style="border: 2px solid orange;" />

Проверим столбец *children*:

In [14]:
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Исправим отрицательные значения "-1", количество элементов незначительно, вероятней всего, это являлось опечаткой и количество относится к элементу "1". Также можно заметить количество детей 20, что смотря на остальные значения, является чем-то неестественным в данной группе,  скорей всего это также является опечаткой и ошибкой при вводе данных, поэтому заменим 20 на 2.

In [15]:
data.loc[data['children'] == -1,'children'] = 1
data.loc[data['children'] == 20,'children'] = 2
data['children'].value_counts() #сделаем проверку

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

**Вывод**

В данном пункте были заменены и проанализированы имеющиеся в столбцах *days_employed* и *total_income* пропуски на среднее значение столбца. После последней проверки видно, что пропусков больше не осталось. Также проверили и отработали столбец *children*, после чего убедились в отсутствии необычных значений.

### Замена типа данных

В данном пункте необходимо провести замену вещественного типа данных на целочисленный. Столбцы *days_employed* и *total_income* имеют тип данных *float* (вещественный), в такой точности нет необходимости, поэтому данные можно преобразовать в тип *int* (целый). Для этого применим метод *astype()*.

In [16]:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**

Провели замену типов данных, после проверки видно, что необходимые столбцы теперь имеют тип *int*.

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>👍</div> 

### Обработка дубликатов

In [17]:
data['education'].value_counts() #посчитали частоту встречаемости уникальных значений

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Можно увидеть, что имеются дубликаты, отличающиеся регистром. Возможная причина - разные менеджера вводили свой вариант значения, из-за чего в итоговом файле они являются дубликатами друг друга. Для исправления этого приведем все строки к нижнему регистру с помощью метода *str.lower()*.

In [18]:
data['education'] = data['education'].str.lower()
data['education'].value_counts() 

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Теперь всё группируется в 5 категорий, и дубликатов нет. Проверим другие столбцы:

In [19]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: income_type, dtype: int64

Дубликатов нет, все 8 групп встречааются в одинаковом виде.

In [20]:
data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

В столбце *purpose* имеются одинаковые цели, выраженные по-разному. Отработаем этот столбец с помощью лемматизации в следующем пункте.

In [21]:
data.duplicated().sum()

71

С помощью функций *duplicated() и sum()* видим, что в таблице одинаковых строк - 71. Удалим их с помощью *drop_duplicates()* и снова проверим таблицу на наличие дубликатов:

In [22]:
data = data.drop_duplicates()

In [23]:
data.duplicated().sum()

0

**Вывод**

В столбце "образование" были обработы дубликаты, в остальных столбцах значения так же были проверены на дубликаты с помощью метода *value_counts()*. После преобразований проверили, имеются ли в таблице дублирующиеся строки. Нашли 71 дубликат и удалили.

<div style="border:solid red 2px; padding: 20px"> <b>Что необходимо доработать</b><br>
После всех преобразований данных необходимо проверить данные на дубликаты</div><br>

<a class="anchor" id="21-bullet">

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Комментарий студента:</b> Не совсем поняла, как еще проверить на дубликаты. Ведь итак использую value_counts в (1) и (2) случае, чтобы показать, что дубликатов нет. </div>   

<div style="border:solid red 2px; padding: 20px"> <b>Что необходимо доработать 2</b><br>
<code>value_counts</code> - применяется к одному столбцу и да его вывод может свидетельствовать о наличии дубликатов. Для проверки на дубликаты (полностью одинаковые строки) есть специальный метод - <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html">pandas.DataFrame.duplicated</a>. Посчитай пожалуйста колв-о дубликатов в получившемся датасете и удали их, если они есть</div><br>

### Лемматизация

Применим лемматизацию для столбца *purpose*.

In [24]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

В данном и предыдущем шаге мы можем видеть уникальные цели по заявкам на кредит, среди них можно выделить основные: операции с жильем, операции по автомобилям, образование, операции с недвижимостью и свадьба. 

В следующем шаге создадим функцию, которая будет лемматизировать все цели, и при нахождении ключевого слова будет выдавать категорию в новом столбце *purpose*. Таким образом, объединим операции с жильем и недвижимостью в единую группу *недвижимость*. Остальные цели сохраним с названиями *свадьба, автомобиль и образование*. Также на случай, если что-то было упущено, добавим опцию *нет категории*.

In [25]:
from pymystem3 import Mystem
m = Mystem()

def create_category(purp):
    purpose = m.lemmatize(purp)
    
    try:
        if 'свадьба' in purpose:
            return 'свадьба'
        if 'автомобиль' in purpose:
            return 'автомобиль'
        if ('жилье' in purpose) or ('недвижимость' in purpose):
            return 'недвижимость'
        if 'образование' in purpose:
            return 'образование'
    except:
        return 'нет категории'
    
data['purposes'] = data['purpose'].apply(create_category)
data['purposes'].value_counts()

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2324
Name: purposes, dtype: int64

**Вывод**

Чтобы лемматизировать столбец *purpose*, для начала получили список уникальных значений из этого столбца методом *unique()*. Затем, лемматизировали каждый элемент из данного списка с помощью написанной функции *create_category*, которая была применена ко всему списку методом *apply()*. Также был создан новый столбец *purposes*, в котором сохранились четыре цели, теперь одинаково сформулированные и которые можно сгруппировать. Также видим, что результат не выдал ничего с опцией *нет категории*, следовательно, всё было сгруппировано верно.

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
Молодец, что формализовала цели👍</div> <br>
<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Лемматизация ресурсно затратный процесс, применять ее к каждой строке очень накладно. Хорошее решение - объединить все укальные цели в строку и к ней применить лемматизацию.</div> <br>
<div style="border:solid red 2px; padding: 20px"> <b>Что необходимо доработать</b><br>
За кадром остался процесс: как ты из всех уникальных целей создала 4. Покажи как ты действовала, даже если это было умозаключение на основе просмотра данных, об этом следует оставить комментарий</div><br>

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа 2</b><br>👍 Спасибо, теперь логика формирования целей понятна</div> 

### Категоризация данных

Для дальнейшего анализа необходимо категоризировать данные. В прошлом шаге был категоризирован столбец с целями. Теперь категоризируем столбец с уровнем дохода и добавим новый столбец *children_count*, в котором количество детей будет категоризировано на три группы.

In [26]:
data['income'] = pd.qcut(data['total_income'],3,labels = ['низкий', 'средний', 'высокий'])
data['income'].value_counts()

средний    7285
низкий     7152
высокий    7017
Name: income, dtype: int64

С помощью функции *qcut* разбили доходы на три равные группы (0.25-, 0.5-, 0.75-квантили): *низкий, средний, высокий*. 

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
Очень классно, что ты знаешь про <code>qcut</code>👍</div> <br>
<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Немного странно выглядит формулировка <code>В данном случае доходы разделились на 0.25-, 0.5-, 0.75-квантили</code> - кажется, что границами разделения являются указанные квартили</div> <br>

In [27]:
def child_count(children):
    if children == 0:
        return 'нет детей'
    elif children == 1 or children == 2:
        return '1-2'
    else: return '3+'

data['children_count'] = data['children'].apply(child_count)
data['children_count'].value_counts()

нет детей    14091
1-2           6983
3+             380
Name: children_count, dtype: int64

Теперь для анализа будет удобней использовать столбец *children_count*, чтобы понять, к какой категории касательно количества детей относится клиент.

<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
<ul>
    <li> категоризация выглядит очень грубой, может можно выделить иные группы, например, <code>нет детей\1-2\3+</code>, как ты думаешь?!
    <li> наименование столбца выглядит тоже не логично - <code>children_id</code> - подразумевает, что у каждого значения в столбце <code>children</code> должен быть какой-то идентификатор. С учетом твоей категоризации - 0\1, лучше использовать <code>is_children</code></ul></div> <br>

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа 2</b><br>👍</div> 

**Вывод**

Для анализа нам нужны будут категоризированные данные уровня дохода, целей кредита, количества детей, а также семейного положения. Столбец с семейным положением уже был отработан, цели кредита были категоризированы, поэтому в этом пункте был рассмотрен и категоризирован столбец с доходом, а именно поделен на три группы, и добавлен новый столбец *children_id*, где хранятся результаты применения функции *child_id*, оценивающей, имеет ли заемщик детей или нет. Если есть хотя бы один ребенок, функция выдает 1, если нет - 0.

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [28]:
data.pivot_table(index=['children_count'],values='debt')

debt
children_count          
1-2             0.092654
3+              0.081579
нет детей       0.075438

In [29]:
data.pivot_table(index=['children_count'],values='debt',aggfunc='sum')

debt
children_count      
1-2              647
3+                31
нет детей       1063

При условии, что детей нет: *P = 1063/14149 = 0.075 = 7.5%*

При условии, что детей в семье 1-2: *P = 647/6996 = 0.092 = 9.2%*

При условии, что детей в семье трое или больше: *P = 31/380 = 0.082 = 8.2%*

**Вывод**

В первой таблице мы видим вероятность существования задолженности при условиях, что у заемщика нет детей, 1-2 детей или 3+ детей. 

Из результата делаем вывод: 
 - вероятность задолженности по сравнению с людьми, у которых нет детей - при наличии одного или двух детей увеличивается на *1.7 процентных пункта до 9.2%*, а при 3+ детей на *0.7 процентных пункта до 8.2%*. 

<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Тут ты оказываешь в заложниках процентов. В твоей формулировке <code>вероятность задолженности при наличии детей увеличивается на 1.7%</code> - говорится о следующем, есть некая вероятность задолжности без детей, например, Х, тогда с детьми она увеличивается на 1.7%, то есть (100% + 1.7%) * Х = 1.017*Х. Подставим твои данные, Х=7.5%, с детьми <code>1.017 * 7.5 = 7.63%</code>, что явно не равно 9.2%. Поэтому корректно говорить о разнице между процентами в процентных пунктах. В данном случае вероятность увеличивает на 1.5 процентных пункта или увеличивается до 9.2%</div> <br>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [30]:
data.pivot_table(index=['family_status'],values='debt').sort_values(by='debt',ascending=False)

debt
family_status                  
Не женат / не замужем  0.097509
гражданский брак       0.093471
женат / замужем        0.075452
в разводе              0.071130
вдовец / вдова         0.065693

**Вывод**

В данной сводной таблице отсортированы в порядке убывания вероятности задолженности при различном семейном статусе. Как мы видим, на первом и втором месте статусы *не женат / не замужем* и *гражданский брак* с вероятностью *9.7% и 9.3%*, затем *женат / замужем, в разводе, вдовец / вдова* с вероятностями *7.5%, 7.1% и 6.6%* соотвественно. Значит, вероятность задолженности по возврату кредита выше примерно на *2.5 процентных пункта*, если человек не состоит или не состоял в официальном браке.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [31]:
data.pivot_table(index=['income'],values='debt').sort_values(by='debt',ascending=False)

debt
income           
средний  0.086754
низкий   0.081236
высокий  0.075246

**Вывод**

Из данной таблицы видно, что задолженности имеются у всех трех групп, при этом вероятность невозврата кредита в срок минимальна у людей с высоким доходом, затем идут заемщики с низким доходом, и наибольшая вероятность у людей со средним уровнем дохода.
У людей со средним уровнем дохода могут быть совершенно разные причины не погашения кредита в срок, поэтому очень сложно определить зависимость между уровнем дохода и погашением кредита.

- Как разные цели кредита влияют на его возврат в срок?

In [32]:
data.pivot_table(index=['purposes'],values='debt').sort_values(by='debt',ascending=False)

debt
purposes              
автомобиль    0.093590
образование   0.092200
свадьба       0.080034
недвижимость  0.072334

**Вывод**

Наибольшую вероятность задолженности имеют люди, взявшие кредит на автомобиль, затем на образование, свадьбу и наименьшую - на недвижимость. 

## Шаг 4. Общий вывод

В проекте был обработан и исследован файл с данными со статистикой о платёжеспособности клиентов. Цель проекта - понять, какие факторы влияют на погашение кредита в срок. Были использованы различные методы обработки данных, группировки, категоризации, чтобы сделать их наиболее "чистыми".

Из проделанного анализа можно сделать вывод о том, какой клиент является наиболее платежеспособным по кредиту и может вернуть его в срок с наибольшей вероятностью. Портрет такого заемщика выглядит так: *человек, не имеющий детей, который состоял или состоит в официальном браке, имеет высокий уровень дохода и цель его кредита - операции с недвижимостью*.

<hr style="border: 2px solid green;" />

## Общий комментарий
Спасибо за проделанную работу🤝

Камиля, ты достаточно хорошо выполнила проект. У нас набралось некоторое кол-во недочетов и ошибок:
- прокомментируй дополнительно лемматизацию
- проверь дубликаты
- формулировки про проценты и их разницу
- кажется, что можно более подробно категоризовать кол-во детей

Исправляй и отправляй на проверку🛠

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.